algviz/interface/visitors.py

```diff
@@ -190,3 +191,59 @@ class WidgetVisitor(Visitor):
 
     def visit(self, *args, **kwargs):
         return super().visit(*args, **kwargs)
+
+@default_for_type(type(None))
+class NullVisitor(Visitor):
+    type_ = Tokens.NULL_T
+
+    def uid(self, item):
+        return structures.Null.uid
+
+    def visit(self, *args, **kwargs):
+        return super().visit(*args, **kwargs)
+
+class TreeVisitor(Visitor):
+    """A visitor for trees of all shapes and sizes"""
+    type_ = Tokens.TREE_NODE_T
+
+    def is_placeholder(self, tree):
+        """Test if the given tree is a placeholder for a non-existent node in
+        rigidly-structured trees.
+
+        For example, in the binary search tree:
+
+            ...........
+            ....2......
+            ..1...3....
+            ........4..
+            ...........
+
+        The first child of `3` should be a placeholder.
+
+        By default, tests if the tree is None.
+        """
+        return tree is None
+
+    @abc.abstractmethod
+    def iter_children(self, tree):
+        """Obviously, return an iterable/generator with the subtrees"""
+        yield from []
+
+    @abc.abstractmethod
+    def get_data(self, tree):
```

Comment:
Why does this just return None? It isn't actually getting any data

Reply:
the `abstractmethod` decorator means that this method isn't a complete implementation.  If you make a subclass of `TreeVisitor` and don't implement `get_data` (and every other abstract method), you won't be able to instantiate the class.
I wrote `return None` as a hint to the reader that `get_data` should return something.  I could, but probably won't, change it to `pass`, which would in fact also return `None`.
```python
>>> from algviz.interface.visitors import TreeVisitor
>>> class Foo(TreeVisitor):
...     pass
...
>>> Foo()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Can't instantiate abstract class Foo with abstract methods get_data, iter_children
```